Introduce dynamic cluster management

pnarayanan opened the pull request:
- Introduce the concept of a cluster participant and a Helix-based
  implementation. The datanodes will register as participants to the local
  Helix/ZK service.
- Introduce a Helix-based dynamic cluster manager that registers to
  Helix/ZK services in all colos as a spectator, coalesces information from the
  different colos, and dynamically updates states based on Helix notifications.
  This will include new implementations for the ClusterMap, DataNodeId,
  PartitionId, DiskId and ReplicaId interfaces.
- Update the failure detector to utilize both Helix-based notifications and
  the old way for detecting failures.
- Rename existing ClusterMap to StaticClusterManager, avoid hardcoding and
  use configs to determine the cluster manager and participant factories to
  use within frontends and datanodes. Make the access specifiers stricter for
  clustermap classes and methods. Other components should only need to deal
  with the cluster map interface.
- Introduce a new metrics class for the Helix-based cluster manager that is separate
  from the static cluster manager metrics.
- To help with the transition from static to dynamic, introduce a
  CompositeClusterManager that wraps over the static and dynamic cluster
  manager instances, uses the static one as the source-of-truth, relays events to
  both, and checks for and reports inconsistencies in the views from the two
  underlying cluster managers.
- Changes were locally tested with multiple ZK servers (simulating multi-colo),
  multiple ambry servers (multi-colo) and one frontend. Regular puts, gets,
  and replication were verified as working successfully - both with the
  HelixClusterManager and the CompositeClusterManager.

Fixes linkedin/ambry#524 and linkedin/ambry#566

--

The changes are a bit expansive due to a few renaming, updates to interfaces and downgrades of public access to package-private for many cluster map classes. However, those should be easy to review. Most of the core code changes are isolated to these files:

New configs, utils:
```
ambry-api/src/main/java/com.github.ambry/config/ClusterMapConfig.java
ambry-clustermap/src/main/java/com.github.ambry.clustermap/ClusterMapUtils.java
```

Cluster participant that registers as a Helix participant to the local colo (only datanodes register as participants):
```
ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixParticipant.java
```

Helix Cluster manager that registers listeners to local and remote colos (both datanodes and frontends register as listners/spectators) :
```
ambry-clustermap/src/main/java/com.github.ambry.clustermap/AmbryDatanode.java
ambry-clustermap/src/main/java/com.github.ambry.clustermap/AmbryDisk.java
ambry-clustermap/src/main/java/com.github.ambry.clustermap/AmbryPartition.java
ambry-clustermap/src/main/java/com.github.ambry.clustermap/AmbryReplica.java
ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java
ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManagerMetrics.java

```

The Composite cluster manager:
```
ambry-clustermap/src/main/java/com.github.ambry.clustermap/CompositeClusterManager.java
```

Updates to the failure detector:
```
ambry-clustermap/src/main/java/com.github.ambry.clustermap/FixedBackoffResourceStatePolicy.java
```

--

All changes were manually tested successfully end-to-end.

Unit test coverage:

```
AmbryDataNode                             100% (1/ 1)    100% (15/ 15)    91.3% (42/ 46)
AmbryDisk                                 100% (1/ 1)    100% (10/ 10)    100% (25/ 25)
AmbryPartition                            100% (1/ 1)    100% (10/ 10)    100% (17/ 17)
AmbryReplica                              100% (1/ 1)    100% (11/ 11)    100% (19/ 19)
ClusterMapUtils                           100% (1/ 1)    88.9% (8/ 9)     92.9% (26/ 28)
CompositeClusterManager                   100% (1/ 1)    100% (10/ 10)    91.8% (67/ 73)
FixedBackoffResourceStatePolicy           100% (1/ 1)    100% (7/ 7)      100% (33/ 33)
FixedBackoffResourceStatePolicyFactory    100% (1/ 1)    100% (2/ 2)      88.9% (8/ 9)
HelixAdminFactory                         100% (1/ 1)    100% (2/ 2)      100% (2/ 2)
HelixBootstrapUpgradeUtil                 100% (1/ 1)    100% (17/ 17)    97.2% (239/ 246)
HelixClusterManager                       100% (5/ 5)    94.9% (37/ 39)   97.1% (231/ 238)
HelixClusterManagerMetrics                100% (16/ 16)  89.5% (34/ 38)   95.6% (86/ 90)
HelixParticipant                          100% (1/ 1)    100% (3/ 3)      100% (20/ 20)
```

vgkholla commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManagerMetrics.java:
```diff
@@ -0,0 +1,231 @@
+  private final HelixClusterManager.HelixClusterManagerCallback clusterMapCallback;
+  private final MetricRegistry registry;
+
+  public Counter liveInstanceChangeTriggerCount;
```
also consider renaming `Manager` as discussed (can be in a future patch if that is easier).

vgkholla commented on ambry-api/src/main/java/com.github.ambry/clustermap/ClusterParticipantFactory.java:
```diff
@@ -0,0 +1,25 @@
+/**
+ * A factory interface to get a {@link ClusterParticipant}
+ */
+public interface ClusterParticipantFactory {
```
reconsider this interface and the factory pattern. As discussed, the type of `ClusterParticipant` is tightly coupled to the the type of `ClusterMap` - may not make sense to have diff factories for them.

vgkholla commented on ambry-api/src/main/java/com.github.ambry/clustermap/ClusterParticipant.java:
```diff
@@ -0,0 +1,30 @@
+  /**
+   * Initialize the participant.
+   */
+  void initialize(String hostname, int port) throws Exception;
```
same comment here - `Exception` feels a bit overbroad.

vgkholla commented on ambry-api/src/main/java/com.github.ambry/clustermap/DiskId.java:
```diff
@@ -24,19 +24,19 @@
    *
    * @return the raw capacity in bytes
    */
-  public long getRawCapacityInBytes();
```
Not sure if you want to make the changes to remove `public` piecemeal. Also seems like it has been done in some places but not elsewhere (for e.g. `ReplicaId`).

vgkholla commented on ambry-api/src/main/java/com.github.ambry/config/ClusterMapConfig.java:
```diff
@@ -66,11 +63,72 @@
+  public final String clusterMapParticipantFactory;
+
+  /**
+   * The clustermap clustermanager factory to use for instantiating the Cluster Manager.
```
used for instantiating the cluster map?

vgkholla commented on ambry-api/src/main/java/com.github.ambry/config/ClusterMapConfig.java:
```diff
@@ -66,11 +63,72 @@
+  /**
+   * The name of the associated datacenter for this node.
+   */
+  @Config("clustermap.datacenter.name")
```
this looks like repeated info since the router has the same info. One of them should no longer have the config (preferably, the router doesn't need to). Otherwise, there is a risk of diverging configs?

Same thing with host name and port.

vgkholla commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/AmbryDatanode.java:
```diff
@@ -0,0 +1,177 @@
+
+
+/**
+ * {@link DataNodeId} implementation to use within dynamic cluster managers.
```
the documentation seems to imply that this can be used with any dynamic cluster manager. Is that true or is this helix/zk focussed?

vgkholla commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/ClusterMapUtils.java:
```diff
@@ -0,0 +1,138 @@
+  static final String ZKCONNECTSTR_STR = "zkConnectStr";
+  static final String ZKINFO_STR = "zkInfo";
+  static final String DATACENTER_STR = "datacenter";
+  static final int MissingRackId = -1;
```
all of these variables in all capitals too?

vgkholla commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/AmbryDatanode.java:
```diff
@@ -0,0 +1,177 @@
+/**
+ * {@link DataNodeId} implementation to use within dynamic cluster managers.
+ */
+class AmbryDataNode extends DataNodeId implements Resource {
```
is there a reason `DataNode` cannot be folded into this i.e. get the static cluster manager to use the same class? Would prefer that since it is gets rid of possibly repeated code.

vgkholla commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/AmbryPartition.java:
```diff
@@ -0,0 +1,116 @@
+  private final HelixClusterManager.HelixClusterManagerCallback clusterManagerCallback;
+  private volatile PartitionState state;
+
+  private static final short Version_Field_Size_In_Bytes = 2;
```
capitals

vgkholla commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/CompositeClusterManager.java:
```diff
@@ -0,0 +1,182 @@
+
+  @Override
+  public PartitionId getPartitionIdFromStream(DataInputStream stream) throws IOException {
+    return staticClusterManager.getPartitionIdFromStream(stream);
```
no comparison required here? Can this fail only due to programming errors and not due to innate differences b/w the two impls?

For e.g., could this fail because a partition exists in the static manager but not in the helix manager? Is that checked on bootstrap (or during load)?

pnarayanan commented on ambry-api/src/main/java/com.github.ambry/clustermap/ClusterAgentsFactory.java:
```diff
@@ -0,0 +1,34 @@
+  /**
+   * Get a reference to a {@link ClusterParticipant}
+   */
+  ClusterParticipant getClusterParticipant() throws IOException;
```
`participant` is a generic enough term in distributed systems and protocols. It is used to represent a constituent entity that has an active part in the workings of the distributed system. Take the [two-phase commit protocol](https://en.wikipedia.org/wiki/Two-phase_commit_protocol), for instance.
I am using the term "agent" to represent any cluster component, including both participants and the cluster manager (which is a listener).

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java:
```diff
@@ -0,0 +1,563 @@
+    private boolean liveInstanceChangeTriggered = false;
+    private boolean externalViewChangeTriggered = false;
+    private boolean configChangeTriggered = false;
+    private CountDownLatch initialized = new CountDownLatch(3);
```
I don't think that is necessary here - it is just internal logic, just like defining three trigger change booleans above.

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java:
```diff
@@ -0,0 +1,563 @@
+    private CountDownLatch initialized = new CountDownLatch(3);
+
+    @Override
+    public void onLiveInstanceChange(List<LiveInstance> liveInstances, NotificationContext changeContext) {
```
Yes, `liveInstanceChange()` provides the list of all live instances at any given time. That is the way Helix does it.

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java:
```diff
@@ -0,0 +1,563 @@
+        liveInstancesSet.add(liveInstance.getInstanceName());
+        instanceNameToDataNodeIdMap.get(liveInstance.getInstanceName()).setState(HardwareState.AVAILABLE);
+      }
+      Set<String> downInstancesSet = Sets.difference(allInstances, liveInstancesSet);
```
No. Fixed.

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java:
```diff
@@ -0,0 +1,563 @@
+      }
+      if (!liveInstanceChangeTriggered) {
+        liveInstanceChangeTriggered = true;
+        initialized.countDown();
```
`initialized` is counted once for each type of change - external view, live instance and config changes.

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java:
```diff
@@ -0,0 +1,563 @@
+  /**
+   * Class that stores all ZK related information associated with a datacenter.
+   */
+  private static class DcZkInfo {
```
It is more of a struct and does not need access to the enclosing parent class. It is a good idea to make it static in such cases.

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java:
```diff
@@ -0,0 +1,563 @@
+     * @param dcName the associated datacenter name.
+     * @param zkConnectStr the associated ZK connect string for this datacenter.
+     * @param helixManager the associated {@link HelixManager} for this datacenter.
+     * @param clusterChangeLIstener the associated {@link ClusterChangeListener} for this datacenter.
```
fixed.

pnarayanan commented on ambry-clustermap/src/test/java/com.github.ambry.clustermap/MockClusterMap.java:
```diff
@@ -188,7 +190,7 @@ public DataNodeId getDataNodeId(String hostname, int port) {
     ArrayList<ReplicaId> replicaIdsToReturn = new ArrayList<ReplicaId>();
     for (PartitionId partitionId : partitions.values()) {
-      List<ReplicaId> replicaIds = partitionId.getReplicaIds();
+      List<? extends ReplicaId> replicaIds = partitionId.getReplicaIds();
```
That is what the interface returns now.

vgkholla commented on ambry-server/src/main/java/com.github.ambry.server/AmbryServer.java:
```diff
@@ -65,28 +67,33 @@
-  public void startup() throws InstantiationException {
+  void startup() throws InstantiationException {
     try {
+      clusterMap = clusterAgentsFactory.getClusterMap();
```
what's the reason to not do this in the constructor?

pnarayanan replied:
Probably wrote it this way because `startup()` seems to do most of the heavy-lifting. And now these involve connecting to a service over the network. `AmbryServer` constructor in fact does not throw (contrary to what the signature says).

vgkholla commented on ambry-tools/src/integration-test/java/com.github.ambry.clustermap/HelixBootstrapUpgradeToolTest.java:
```diff
@@ -92,6 +50,8 @@ void shutdown() {
   private final JSONObject zkJson;
   private TestHardwareLayout testHardwareLayout;
   private TestPartitionLayout testPartitionLayout;
+  private final String clusterNameInStaticClusterMap = "AmbryToolTestStatic";
```
could this be a static constant?

pnarayanan replied:
yes, fixed.

vgkholla commented on config/server.properties:
```diff
@@ -11,3 +11,4 @@
 #
 
 host.name=localhost
+clustermap.cluster.name=Ambry_Dev
```
it's ok for a cluster to not exist?

pnarayanan replied:
for dev setup, we don't currently handle Helix based instantiation. This can be added later. For dev testing, we need to use the configs named with 'helix' in them. Will add a write-up at some point on the steps.

vgkholla commented on config/log4j.properties:
```diff
@@ -26,5 +26,5 @@ log4j.appender.PublicAccessLog.File=${ambry.log.dir}/publicAccessLog.out
 log4j.appender.PublicAccessLog.layout.conversionPattern=%m%n
 
-log4j.logger.PublicAccessLogger = INFO, PublicAccessLog
+log4j.logger.PublicAccessLogger = ERROR, PublicAccessLog
```
why is this required?

pnarayanan replied:
INFO logging is generally unnecessary, and when needed may be enabled. The logs can otherwise grow considerably if a cluster with multiple AmbryServers are instantiated due to replication requests and responses filling them up.

vgkholla commented on config/server6_helix.properties:
```diff
@@ -0,0 +1,20 @@
+#
```
are these many property files required?

pnarayanan replied:
Yes - these are helpful in proper simulation.

vgkholla commented:
Some of my comments have been collapsed because I was looking at an older copy.

vgkholla commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java:
```diff
@@ -0,0 +1,563 @@
+   * @param instanceConfig the {@link InstanceConfig} associated with this datanode.
+   * @throws Exception if creation of {@link AmbryDisk} throws an Exception.
+   */
+  private void initializeDisksAndReplicasOnNode(AmbryDataNode datanode, InstanceConfig instanceConfig)
```
It's true that all the information can be gathered from the code but it would be nice to have it documented what an `InstanceConfig` contains (or is expected to contain) for easy reference.

xiahome commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java:
```diff
@@ -0,0 +1,563 @@
+      AmbryDisk disk = new AmbryDisk(clusterMapConfig, datanode, mountPath, state, capacityBytes);
+      dataNodeIdToDiskIds.get(datanode).add(disk);
+
+      if (!replicasStr.isEmpty()) {
```
Should it be `replicasStr != null`?

pnarayanan replied:
No `replicasStr` will always be non-null actually. It just could be empty.

xiahome commented on ambry-api/src/main/java/com.github.ambry/clustermap/ClusterMap.java:
```diff
@@ -26,23 +26,23 @@
    * @return deserialized PartitionId
    */
-  public PartitionId getPartitionIdFromStream(DataInputStream stream) throws IOException;
+  PartitionId getPartitionIdFromStream(InputStream stream) throws IOException;
```
I guess my question is why it could not be moved to `ClusterMapUtils`?

pnarayanan replied:
Because this is implementation dependent. Internally they may (and do) use a common method in `ClusterMapUtils`.

xiahome commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/CompositeClusterAgentsFactory.java:
```diff
@@ -0,0 +1,91 @@
+      try {
+        helixClusterManager = helixClusterAgentsFactory.getClusterMap();
+      } catch (Exception e) {
+        logger.error("Helix cluster manager instantiation failed with exception", e);
```
@vgkholla It seems there is a path in `CompositeClusterManager` to allow `null` for `HelixClusterManager`.

pnarayanan replied:
There is a metric within the `HelixClusterManager`. We could fail, but this way is just a little safer during deployments the way I imagined doing them (a little more confidence in enabling on multiple nodes at the same time).

xiahome commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/Datacenter.java:
```diff
@@ -38,7 +38,7 @@
   private Logger logger = LoggerFactory.getLogger(getClass());
 
-  public Datacenter(HardwareLayout hardwareLayout, JSONObject jsonObject, ClusterMapConfig clusterMapConfig)
+  Datacenter(HardwareLayout hardwareLayout, JSONObject jsonObject, ClusterMapConfig clusterMapConfig)
```
Could you add javadoc for this method? It is not very clear what kind information the `JSONObject` would contain. Also, it is a bit strange a DC is instantiated with its key (i.e., its name) hidden in a `JSONObject`.

pnarayanan replied:
You could just ignore this, these are static cluster manager specific (added the doc). 
I did not make any significant changes to this class (just changed the access specifiers). Helix cluster manager will no longer use this class.

xiahome commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/Disk.java:
```diff
@@ -26,19 +26,15 @@
-  private static final long MinCapacityInBytes = 10L * 1024 * 1024 * 1024;
-  private static final long MaxCapacityInBytes = 10L * 1024 * 1024 * 1024 * 1024; // 10 TB
-
+class Disk implements DiskId {
```
Could you add in javadoc that this class is used for static cluster management?

pnarayanan replied:
Added.

xiahome commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java:
```diff
@@ -0,0 +1,573 @@
+        dcZkInfo.helixManager.addExternalViewChangeListener(dcZkInfo.clusterChangeListener);
+        dcZkInfo.helixManager.addConfigChangeListener(dcZkInfo.clusterChangeListener);
+        dcZkInfo.helixManager.addLiveInstanceChangeListener(dcZkInfo.clusterChangeListener);
+        dcZkInfo.clusterChangeListener.waitForInitialization();
```
If `HelixClusterManager` is used in the `CompositeClusterManager`, is it necessary to wait until the `HelixClusterManager` to receive information from all DCs? Should the `StaticClusterManager` be used until the the helix-based one is up? It seems to be one of the benefits to use the `CompositeClusterManager`, as you do not have to depend on one completely.

pnarayanan replied:
I don't follow your question completely - but the `CompositeClusterManager` will be used during the transition from static to dynamic. Going forward, we will only use the `HelixClusterManager` and get rid of the static cluster manager.

xiahome commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixParticipant.java:
```diff
@@ -0,0 +1,78 @@
+/**
+ * An implementation of {@link ClusterParticipant} that registers registers as a participant to a Helix cluster.
+ */
+class HelixParticipant implements ClusterParticipant {
```
One question I am not completely sure about is, how a `HelixParticipant` participates in a cluster. The logic implemented in this class seem not to report changes at a granular basis. Maybe I missed some meaningful parts in the code, so I will come to you next week for more clarity.

pnarayanan replied:
I think I answered this as part of a previous comment of yours - let us discuss this in person if you need more clarity.

xiahome commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/Partition.java:
```diff
@@ -27,18 +27,18 @@
  * consist of one or more {@link Replica}s. Replicas ensure that a Partition is available and reliable.
  */
-public class Partition extends PartitionId {
+class Partition extends PartitionId {
```
Add in javadoc saying this is for static cluster management?

pnarayanan replied:
Yes. Added docs for `Partition`, `Replica`, `Disk`,`DataNode` and `DataCenter` - all static cluster manager related classes.

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java:
```diff
@@ -0,0 +1,552 @@
+  private final Map<AmbryPartition, Set<AmbryReplica>> partitionIdToReplicaIds = new ConcurrentHashMap<>();
+  private final Map<AmbryDataNode, Set<AmbryReplica>> dataNodeIdToReplicaIds = new ConcurrentHashMap<>();
+  private final Map<AmbryDataNode, Set<AmbryDisk>> dataNodeIdToDiskIds = new ConcurrentHashMap<>();
+  private final Map<ByteBuffer, AmbryPartition> partitionMap = new ConcurrentHashMap<>();
```
For the running process? I don't think it would be prohibitively high. These mappings would one way or the other already exist today in the static cluster manager.

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java:
```diff
@@ -0,0 +1,563 @@
+      for (String instanceName : downInstancesSet) {
+        instanceNameToDataNodeIdMap.get(instanceName).setState(HardwareState.UNAVAILABLE);
+      }
+      if (!liveInstanceChangeTriggered) {
```
That is what I was told - but have now added explicit synchronization between so as to guarantee that all the notification methods are serialized - as Helix documentation does not make it clear. If the caller indeed serializes the notifications, then there wouldn't be any contention and the lock would be harmless.

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java:
```diff
@@ -0,0 +1,563 @@
+    @Override
+    public void onConfigChange(List<InstanceConfig> configs, NotificationContext changeContext) {
+      logger.info("Config change triggered with: ", configs);
+      super.onConfigChange(configs, changeContext);
```
Super class did not implement a `LiveInstanceChangeListener` interestingly. Anyway, based on @xiahome's comment, we now implement these listeners without extending the `RoutingTableProvider`.

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManagerMetrics.java:
```diff
@@ -0,0 +1,231 @@
+    Gauge<Long> datacenterCount = new Gauge<Long>() {
+      @Override
+      public Long getValue() {
+        return clusterMapCallback.getDatacenterCount();
```
This is carried over from the static cluster manager, where this could just have been initialized once, really. With the dynamic cluster manager, one could introduce a new datacenter dynamically without recycling existing instances (which would happen only rarely, of course). So this, interestingly, has some relevance now.

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/ClusterMapUtils.java:
```diff
@@ -0,0 +1,138 @@
+   * @param instanceConfig the {@link InstanceConfig} associated with the interested instance.
+   * @return the list of sealed replicas.
+   */
+  static List<String> getSealedReplicas(InstanceConfig instanceConfig) {
```
I think they do - this class's contract does not say that it is intended to be generic. The practical use is to keep utility methods that are useful more than once. Only requirement is that the method should be static.

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/StaticClusterAgentsFactory.java:
```diff
@@ -0,0 +1,88 @@
+  @Override
+  public ClusterParticipant getClusterParticipant() throws IOException {
+    if (clusterMapConfig == null) {
+      throw new IllegalArgumentException(
```
For tests that know that they do not need to instantiate a participant, and instantiate the factory via the method that does not really use the `ClusterMapConfig`. Just a bit easier to write tests.

pnarayanan replied:
This isn't really required anymore - so have removed it.

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java:
```diff
@@ -0,0 +1,573 @@
+    try {
+      Map<String, String> dataCenterToZkAddress =
+          parseZkJsonAndPopulateZkInfo(clusterMapConfig.clusterMapDcsZkConnectStrings);
+      for (Map.Entry<String, String> entry : dataCenterToZkAddress.entrySet()) {
```
Yes. Helix does not yet have the functionality of relaying cross-colo notifications.

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java:
```diff
@@ -0,0 +1,573 @@
+      initialize();
+      // Now register listeners to get notified on change.
+      for (DcZkInfo dcZkInfo : dcToDcZkInfo.values()) {
+        dcZkInfo.helixManager.addExternalViewChangeListener(dcZkInfo.clusterChangeListener);
```
We are not currently using it, but this notification will come once per resource when nodes go down as I understand it, and it will contain the partitions and their instance mappings and so on. We will make use of this in the future.

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java:
```diff
@@ -0,0 +1,573 @@
+  /**
+   * An instance of this object is used to register as listener for Helix related changes in each datacenter.
+   */
+  private class ClusterChangeListener implements ExternalViewChangeListener, ConfigChangeListener, LiveInstanceChangeListener {
```
Yes - have replaced it with `InstanceConfigChangeListener` now.
(previously had to use the deprecated one, because `RoutingTableProvider` still used it and for some reason registering as an `InstanceConfigChangeListener` did not work out - no notifications on changes).

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/CompositeClusterAgentsFactory.java:
```diff
@@ -0,0 +1,91 @@
+public class CompositeClusterAgentsFactory implements ClusterAgentsFactory {
+  private static final Logger logger = LoggerFactory.getLogger(CompositeClusterAgentsFactory.class);
+  private final StaticClusterAgentsFactory staticClusterAgentsFactory;
+  private final HelixClusterAgentsFactory helixClusterAgentsFactory;
```
It does not strictly have to - but at this time, I don't see a reason to make these too generic. If we see value in keeping the composite cluster manager after transitioning to Helix, we can think of something.

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/ClusterMapUtils.java:
```diff
@@ -0,0 +1,154 @@
+  static Map<String, String> parseZkJsonAndPopulateZkInfo(String zkLayoutJsonString) throws JSONException {
+    Map<String, String> dataCenterToZkAddress = new HashMap<>();
+    JSONObject root = new JSONObject(zkLayoutJsonString);
+    JSONArray all = (root).getJSONArray(ZKINFO_STR);
```
Not needed. I must have at one point had it as
`(new JSONObject(zkLayoutJsonString)).getJSONArray()`.

pnarayanan commented on ambry-server/src/main/java/com.github.ambry.server/AmbryServer.java:
```diff
@@ -143,6 +150,8 @@ public void startup() throws InstantiationException {
           networkServer.getRequestResponseChannel(), requests);
       networkServer.start();
 
+      clusterParticipant.initialize(networkConfig.hostName, networkConfig.port);
```
Participant has a corresponding `Instance` in Helix. When participant disconnects, Helix considers the instance as down and ensures that some other instance is the LEADER for all the partitions for which the down instance was the LEADER. Going forward, when the Ambry server detects disk failures, it will simply update this information in its `InstanceConfig`.

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/CompositeClusterManager.java:
```diff
@@ -0,0 +1,272 @@
+        staticPartitionIdStrings.add(partitionId.toString());
+      }
+
+      Set<String> helixPartitionIdStrings = new HashSet<>();
```
This verification will never be removed, as the purpose of this class is to do this verification and update the metrics. We will at some point move to using the HelixClusterManager.

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/CompositeClusterManager.java:
```diff
@@ -0,0 +1,272 @@
+      }
+
+      if (!staticPartitionIdStrings.equals(helixPartitionIdStrings)) {
+        helixClusterManagerMetrics.getWritablePartitionIdsMismatchCount.inc();
```
Going forward, nodes will update their `InstanceConfigs`. Right now, all changes are done by updating the static files and running the bootstrap utility.

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/AmbryStateModel.java:
```diff
@@ -0,0 +1,68 @@
+ * {@link StateModel} to use when the Ambry participants (datanodes) register to Helix. The methods are callbacks
+ * that get called within a participant whenever its state changes in Helix. For now these are no-ops.
+ *
+ * In the criticial path of puts and gets, there are no leader replicas in Ambry. Every replica can serve reads and
```
fixed, thanks.

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/AmbryStateModel.java:
```diff
@@ -0,0 +1,68 @@
+
+  @Transition(to = "STANDBY", from = "OFFLINE")
+  public void onBecomeStandbyFromOffline(Message message, NotificationContext context) {
+    logger.info("Becoming STANDBY from OFFLINE");
```
STANDBY *from* OFFLINE. That information must be retrievable from the `Message`. Have not explored it in detail yet.

pnarayanan commented:
@vgkholla @xiahome Addressed your comments, added unit tests and updated the coverage in the description. Please take a look at the last two commits for the newly added tests and changes.

vgkholla commented on ambry-clustermap/src/test/java/com.github.ambry.clustermap/FixedBackoffResourceStatePolicyTest.java:
```diff
@@ -0,0 +1,115 @@
+
+
+/**
+ *
```
empty

pnarayanan replied:
this was added in the commit after this.

vgkholla commented on ambry-clustermap/src/test/java/com.github.ambry.clustermap/FixedBackoffResourceStatePolicyTest.java:
```diff
@@ -0,0 +1,115 @@
+  }
+
+  @Test
+  public void testThresholdsAndRetries() {
```
doc for both

pnarayanan replied:
this was also added in the later commit.

vgkholla commented on ambry-clustermap/src/test/java/com.github.ambry.clustermap/FixedBackoffResourceStatePolicyTest.java:
```diff
@@ -0,0 +1,115 @@
+    time.sleep(RETRY_BACKOFF_MS - RETRY_BACKOFF_MS / 2);
+    assertFalse(policy.isDown());
+
+    // At this time the resource is conditionally up. Verify that a single error should bring it back down.
```
is there a way to verify that it is actually conditionally up?

pnarayanan replied:
The fact that it is not down and the verification that a single subsequent error brings it down is in fact testing that it was conditionally up.

pnarayanan commented on ambry-api/src/main/java/com.github.ambry/clustermap/ClusterMap.java:
```diff
@@ -79,4 +79,9 @@
+  /**
+   * Close the cluster map. Any cleanups should be done in this call.
+   */
+  void close();
```
That is a good idea. As discussed, will have both `ClusterMap` and `ClusterParticipant` implement `AutoCloseable` and override `close()` in the respective interfaces to not throw exceptions.

vgkholla commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterAgentsFactory.java:
```diff
@@ -0,0 +1,68 @@
+    this(clusterMapConfig, new MetricRegistry());
+  }
+
+  public HelixClusterAgentsFactory(ClusterMapConfig clusterMapConfig, MetricRegistry metricRegistry) {
```
why are there two public constructors?

pnarayanan replied:
This does not have to be public anymore.

vgkholla commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManagerMetrics.java:
```diff
@@ -0,0 +1,242 @@
+  }
+
+  void initializeInstantiationMetric(final boolean instantiated) {
+    helixClusterManagerInstantiationFailed = new Gauge<Boolean>() {
```
why is this a gauge? Once a value is set, this can never change right? Counter seems better.

pnarayanan replied:
A boolean gauge is better than a counter for this I feel. Why have a counter when it can only be updated once?

vgkholla commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixParticipant.java:
```diff
@@ -0,0 +1,85 @@
+      throw new IllegalStateException("Clustername is empty in clusterMapConfig");
+    }
+    try {
+      zkConnectStr = ClusterMapUtils.parseZkJsonAndPopulateZkInfo(clusterMapConfig.clusterMapDcsZkConnectStrings)
```
could this be passed in?

pnarayanan replied:
Probably. Have added this to the backlog to see if we can nicely pass this for the `HelixParticipant` and the entire map for the `HelixClusterManager`. There are cascading changes (different exceptions thrown that have to be either caught or thrown with an update to the interface method signature).

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java:
```diff
@@ -344,49 +356,55 @@ AmbryReplica getReplicaForPartitionOnNode(String hostname, int port, String part
-        } else {
-          instanceNameToDataNodeIdMap.get(instanceName).setState(HardwareState.UNAVAILABLE);
+      synchronized (notificationLock) {
+        logger.info("Live instance change triggered with: " + liveInstances);
```
Wanted this as INFO only for the transitioning phase for studying the patterns. TRACE/DEBUG comes with excessive logging that is unnecessary unless chasing an issue. That said, I have changed this to TRACE as this consists of all instances, not just the changed ones (as I had thought originally).

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java:
```diff
@@ -427,7 +445,7 @@ void waitForInitialization() throws InterruptedException {
    * A callback class used to query information from the {@link HelixClusterManager}
    */
-  class HelixClusterManagerCallback {
+  public class HelixClusterManagerCallback {
```
Does not. Changed. (It is an inner class of a package private class, so it was not really public).

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixFactory.java:
```diff
@@ -0,0 +1,38 @@
+/**
+ * A factory class to construct and get a reference to a {@link HelixManager}
+ */
+public class HelixFactory {
```
Prefer it this way as there is a `HelixManagerFactory` in Helix and want to avoid confusion.

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixFactory.java:
```diff
@@ -0,0 +1,38 @@
+   */
+  HelixManager getZKHelixManager(String clusterName, String instanceName, InstanceType instanceType,
+      String zkAddr) {
+    return HelixManagerFactory.getZKHelixManager(clusterName, instanceName, instanceType, zkAddr);
```
This was added to help with the tests. The `getZKHelixManager` method in the Helix class `HelixManagerFactory` is unfortunately static and cannot be overridden.

pnarayanan commented on ambry-rest/src/main/java/com.github.ambry.rest/RestServer.java:
```diff
@@ -280,8 +284,14 @@ public void shutdown() {
       logger.info("Router close took {} ms", elapsedTime);
       restServerMetrics.routerCloseTime.update(elapsedTime);
 
+      clusterMap.close();
```
Fair point. Added this to the finally of the try block in which the `clusterMap` is instantiated.

pnarayanan commented on ambry-server/src/main/java/com.github.ambry.server/AmbryServer.java:
```diff
@@ -216,6 +216,7 @@ void shutdown() {
           logger.error("Error while closing notification system.", e);
         }
       }
+      clusterMap.close();
```
Didn't realize that `shutdown()` could be called even if `startup()` did not succeed (which is a bit strange to begin with - ideally startup() should cleanup before returning a failure). I have added documentation for the shutdown methods in both the `RestServer` and `AmbryServer` to reflect this fact. Have also added the `null` checks.

And the factory creates agents only when asked for. This is particularly important with the participants (which will not be created by the frontends). The factories are also singletons - that is they only construct each type of agent, once. I have added this to the `ClusterAgentsFactory` interface.

pnarayanan commented on ambry-clustermap/src/test/java/com.github.ambry.clustermap/HelixClusterManagerTest.java:
```diff
@@ -0,0 +1,619 @@
+          metricRegistry);
+      fail("Instantiation should have failed with invalid zk addresses");
+    } catch (IOException e) {
+      assertEquals(true, metricRegistry.getGauges()
```
`assertTrue()` requires a cast as the second argument is generic at compile time. Messages are unnecessary in these cases (should be clear from the code that instantiation failure metric was expected to be true but was false actually). Better to keep it less verbose for readability.

pnarayanan commented on ambry-clustermap/src/test/java/com.github.ambry.clustermap/HelixClusterManagerTest.java:
```diff
@@ -0,0 +1,619 @@
+    props.setProperty("clustermap.datacenter.name", "DC1");
+    ClusterMapConfig clusterMapConfig2 = new ClusterMapConfig(new VerifiableProperties(props));
+
+    // AmbryDataNode test
```
The cluster components are better tested together - as they are interrelated - need to create `AmbryDataNode` for creating an `AmbryDisk`, for example. Having them all in a single method makes things easier.

Have moved this whole method to a test of its own, independent of Helix, as this is intended to be used with *any* dynamic cluster manager. Have also made `HelixClusterManagerCallback` and implementation of an interface called `ClusterManagerCallback`, which is what the dynamic components will use.

pnarayanan commented on ambry-api/src/main/java/com.github.ambry/config/ClusterMapConfig.java:
```diff
@@ -25,11 +25,8 @@
-   * number of consecutive errors, and if so, considers the resource as down for
-   * 'retry backoff' milliseconds.
-   */
+  // The fixed timeout based resource state handling checks if we have had a 'threshold' number of consecutive errors,
```
Removed, as this is unnecessary as you say - the documentation is there in the implementation.

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/AmbryDisk.java:
```diff
@@ -0,0 +1,121 @@
+    this.datanode = datanode;
+    this.mountPath = mountPath;
+    this.rawCapacityBytes = rawCapacityBytes;
+    ResourceStatePolicyFactory resourceStatePolicyFactory =
```
Actually this can't be passed in as is (contrary to what I led you to believe) as `AmbryDisk` and `AmbryDatanode` implement the `Resource` interface and the factory needs that resource.

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/AmbryPartition.java:
```diff
@@ -0,0 +1,117 @@
+
+  @Override
+  public byte[] getBytes() {
+    ByteBuffer buffer = ByteBuffer.allocate(PARTITION_SIZE_IN_BYTES);
```
Moved this to a common method in `ClusterMapUtils`. This is just serializing a short and a long.

pnarayanan commented:
@vgkholla @xiahome Addressed comments and updated.

vgkholla commented on ambry-rest/src/main/java/com.github.ambry.rest/RestServer.java:
```diff
@@ -93,6 +94,7 @@
     public final Histogram restServerShutdownTimeInMs;
     public final Histogram restServerStartTimeInMs;
     public final Histogram routerCloseTime;
+    public final Histogram clusterMapCloseTime;
```
not needed anymore.

pnarayanan replied:
Fixed.

vgkholla commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java:
```diff
@@ -0,0 +1,592 @@
+    dataNodeIdToReplicaIds.put(datanode, new HashSet<AmbryReplica>());
+    dataNodeIdToDiskIds.put(datanode, new HashSet<AmbryDisk>());
+    List<String> sealedReplicas = getSealedReplicas(instanceConfig);
+    Map<String, Map<String, String>> diskInfos = instanceConfig.getRecord().getMapFields();
```
could you explain how this works? What does `getMapFields()` do? Are all the map fields only disk info? What if we add a map field in the future that isn't disk info?

pnarayanan replied:
That is a good question. I found Helix a bit inflexible with regard to this. It is a conscious decision to store diskInfos in the map fields. This helps in making this readable externally (rather than serializing information, say). The downside is that the map fields cannot store anything else - which is okay for now and for the foreseeable future.
I have an item in the backlog to making this extensible. The simplest thing is to add a version as a simple field. I just want to think of that a bit more generally.

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java:
```diff
@@ -0,0 +1,592 @@
+  private long clusterWideAllocatedRawCapacityBytes;
+  private long clusterWiseAllocatedUsableCapacityBytes;
+  private final HelixClusterManagerCallback helixClusterManagerCallback;
+  final HelixClusterManagerMetrics helixClusterManagerMetrics;
```
`CompositeClusterManager` needs access to this.

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java:
```diff
@@ -0,0 +1,592 @@
+  private final MetricRegistry metricRegistry;
+  private final ClusterMapConfig clusterMapConfig;
+  private final Map<String, DcZkInfo> dcToDcZkInfo = new HashMap<>();
+  private final Map<String, AmbryPartition> partitionNameToPartitionIdMap = new ConcurrentHashMap<>();
```
Thanks, changed.

pnarayanan commented:
@vgkholla @xiahome Updated. Also added documentation on what we store in the `InstanceConfig` in this patch itself. See the javadoc for `HelixBootstrapUpgradeUtil` class.
